# Mutation Rate
Reference for the paper:

## Spontaneous mutation rate estimation of the principal malaria vector Anopheles coluzzii and Anopheles stephensi
Iliyas Rashid, Melina Campos, Travis Collier, Marc Crepeau, Allison Weakley, Hans Gripkey, Yoosook Lee, Hanno Schmidt, Gregory C. Lanzaro

We estimated the nuclear mutation rate per generation in the malaria vectors Anopheles coluzzii and Anopheles stephensi by detecting de novo genetic mutations.

### Input file info (VCF file)
VCF file generated by Freebayes was named pconcat.vcf.gz.  This file was then typically manipulated with the following command to rearrange the sample order:

`vcf-shuffle-cols -t shuffletemplate.vcf pconcat.vcf.gz > pconcat_shuffle.vcf`

Then the data was filtered with this command to retain only biallelic SNPs:

`bcftools view -O z -m2 -M2 -v snps -o pconcat_biallelic_shuffle.vcf.gz pconcat_shuffle.vcf`

Once the correct father was identified, data for the incorrect fathers was removed to create an input file for Iliyas’s mutation detection program.  This input file is named pconcat-order-pass.vcf.gz


### Male parent identification
-	Relatedness (--relatedness2 from VCFtools):

`vcftools --gzvcf pconcat_biallelic_shuffle.vcf.gz –relatedness2 --out pconcat_biallelic_shuffle`
  
-	Mendelian heritage: 
Perl script `Mendelian_Inheritance_Analysis.pl`

### De novo mutation detection
-	De novo mutation detection 
Perl script `DenovoMutationsDetection.pl”` : this is the main program for de novo mutation detection

### Synthetic mutation insertion
-	Distribution of real data alternate allele depth;
-	Script in Python from Travis
